File: src/blockchain.py
import hashlib
import json
import time
from typing import List, Dict, Any

class Block:
    def __init__(self, index: int, timestamp: float, data: Any, previous_hash: str, proof: int = 0):
        self.index = index
        self.timestamp = timestamp
        self.data = data
        self.previous_hash = previous_hash
        self.proof = proof
        self.hash = self.calculate_hash()

    def calculate_hash(self) -> str:
        block_string = json.dumps({
            "index": self.index,
            "timestamp": self.timestamp,
            "data": self.data,
            "previous_hash": self.previous_hash,
            "proof": self.proof
        }, sort_keys=True).encode()
        return hashlib.sha256(block_string).hexdigest()

    def __repr__(self):
        return f"Block(index={self.index}, hash={self.hash[:8]}, prev={self.previous_hash[:8]})"

class Blockchain:
    def __init__(self, difficulty: int = 4):
        self.chain: List[Block] = []
        self.difficulty = difficulty
        self.create_genesis_block()

    def create_genesis_block(self):
        genesis_block = Block(0, time.time(), "Genesis Block", "0")
        self.chain.append(genesis_block)

    @property
    def last_block(self) -> Block:
        return self.chain[-1]

    def add_block(self, block: Block) -> bool:
        if self.is_valid_new_block(block, self.last_block):
            self.chain.append(block)
            return True
        return False

    def proof_of_work(self, index: int, timestamp: float, data: Any, previous_hash: str) -> int:
        proof = 0
        while True:
            hash_val = self._calculate_hash_with_proof(index, timestamp, data, previous_hash, proof)
            if hash_val.startswith('0' * self.difficulty):
                return proof
            proof += 1

    def _calculate_hash_with_proof(self, index: int, timestamp: float, data: Any, previous_hash: str, proof: int) -> str:
        block_string = json.dumps({
            "index": index,
            "timestamp": timestamp,
            "data": data,
            "previous_hash": previous_hash,
            "proof": proof
        }, sort_keys=True).encode()
        return hashlib.sha256(block_string).hexdigest()

    def is_valid_new_block(self, new_block: Block, previous_block: Block) -> bool:
        if previous_block.index + 1 != new_block.index:
            return False
        if previous_block.hash != new_block.previous_hash:
            return False
        if new_block.hash != new_block.calculate_hash():
            return False
        if not new_block.hash.startswith('0' * self.difficulty):
            return False
        return True

    def is_chain_valid(self, chain: List[Block]) -> bool:
        for i in range(1, len(chain)):
            if not self.is_valid_new_block(chain[i], chain[i-1]):
                return False
        return True

    def to_dict(self):
        return [vars(b) for b in self.chain]

File: src/pkg/blockchain.go
package pkg

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
	"time"
)

type LogData struct {
	Event     string `json:"event"`
	PublicKey string `json:"public_key"`
	Signature string `json:"signature"`
}

type Block struct {
	Index        int       `json:"index"`
	Timestamp    int64     `json:"timestamp"`
	Data         interface{} `json:"data"`
	PreviousHash string    `json:"previous_hash"`
	Proof        int       `json:"proof"`
	Hash         string    `json:"hash"`
}

func (b *Block) CalculateHash() string {
	record := fmt.Sprintf("%d%d%v%s%d", b.Index, b.Timestamp, b.Data, b.PreviousHash, b.Proof)
	h := sha256.New()
	h.Write([]byte(record))
	hashed := h.Sum(nil)
	return hex.EncodeToString(hashed)
}

type Blockchain struct {
	Chain      []Block
	Difficulty int
}

func NewBlockchain(difficulty int) *Blockchain {
	bc := &Blockchain{
		Difficulty: difficulty,
	}
	bc.createGenesisBlock()
	return bc
}

func (bc *Blockchain) createGenesisBlock() {
	genesis := Block{
		Index:        0,
		Timestamp:    time.Now().Unix(),
		Data:         "Genesis Block",
		PreviousHash: "0",
		Proof:        0,
	}
	genesis.Hash = genesis.CalculateHash()
	bc.Chain = append(bc.Chain, genesis)
}

func (bc *Blockchain) GetLastBlock() Block {
	return bc.Chain[len(bc.Chain)-1]
}

func (bc *Blockchain) IsValidNewBlock(newBlock Block, previousBlock Block) bool {
	if previousBlock.Index+1 != newBlock.Index {
		return false
	}
	if previousBlock.Hash != newBlock.PreviousHash {
		return false
	}
	if newBlock.Hash != newBlock.CalculateHash() {
		return false
	}
	if !strings.HasPrefix(newBlock.Hash, strings.Repeat("0", bc.Difficulty)) {
		return false
	}
	return true
}

func (bc *Blockchain) ProofOfWork(index int, data interface{}, prevHash string) (int, int64) {
	proof := 0
	timestamp := time.Now().Unix()
	for {
		record := fmt.Sprintf("%d%d%v%s%d", index, timestamp, data, prevHash, proof)
		h := sha256.New()
		h.Write([]byte(record))
		hash := hex.EncodeToString(h.Sum(nil))
		if strings.HasPrefix(hash, strings.Repeat("0", bc.Difficulty)) {
			return proof, timestamp
		}
		proof++
	}
}

func (bc *Blockchain) AddBlock(newBlock Block) bool {
	if bc.IsValidNewBlock(newBlock, bc.GetLastBlock()) {
		bc.Chain = append(bc.Chain, newBlock)
		return true
	}
	return false
}

func (bc *Blockchain) IsChainValid(chain []Block) bool {
	if len(chain) == 0 {
		return false
	}
	// Simplified validation for this version
	for i := 1; i < len(chain); i++ {
		if !bc.IsValidNewBlock(chain[i], chain[i-1]) {
			return false
		}
	}
	return true
}
